### DISCUSSION TITLE
Postgresql setup enum mapping for the particular database

### DISCUSSION THREAD

**Discussion by ademchenko** [linq2db/linq2db#3727] [Q&A] [answered]
At the moment we are using linq2db 3.7.0. We are following #3453 to map Postgres enums to our .net enums. 
The advice in the thread is to register enum type globally for npgsql:
`NpgsqlConnection.GlobalTypeMapper.MapEnum<TimeUnit>("time_unit");`

But the thing is that we have two databases and would like to register that mapping just for one of them. 
According to the [docs of Npgsql](https://www.npgsql.org/doc/types/enums_and_composites.html), it's possible to achieve that on a per-connection basis:
```
var conn = new NpgsqlConnection(...);
conn.TypeMapper.MapEnum<SomeEnum>("some_enum_type");
conn.TypeMapper.MapComposite<SomeType>("some_composite_type");
```
How could we do that in linq2db where creating of the connection is hidden from us? Even further, It would not be convenient if at every place of using our first but not the second database we would need to configure the mappings (like in the example above).
Is there a way to configure a mapping on a per-database level (so, once for a particular database) having that we set up our linq2db settings in the following way:

```
DataConnection.DefaultSettings = new LinqToDbSettings(
                new ConnectionStringSettings(Configuration.GetConnectionString("FirstDatabaseConnectionName"), ProviderName.PostgreSQL, "FirstDatabaseConnectionName"),
                new ConnectionStringSettings(Configuration.GetConnectionString("SecondDatabaseConnectionName"), ProviderName.PostgreSQL, "SecondDatabaseConnectionName"));

public class LinqToDbSettings : ILinqToDBSettings
    {
        public IEnumerable<IDataProviderSettings> DataProviders => Enumerable.Empty<IDataProviderSettings>();

        public string DefaultConfiguration { get; }

        public string DefaultDataProvider { get; }

        public IEnumerable<IConnectionStringSettings> ConnectionStrings { get; }

        public LinqToDbSettings(string connectionName, string dataProvider, string connectionString)
        {
            if (connectionName == null) throw new ArgumentNullException(nameof(connectionName));
            if (dataProvider == null) throw new ArgumentNullException(nameof(dataProvider));
            if (connectionString == null) throw new ArgumentNullException(nameof(connectionString));

            DefaultConfiguration = dataProvider;
            DefaultDataProvider = dataProvider;
            ConnectionStrings = Enumerable.Repeat(new ConnectionStringSettings(connectionString, dataProvider, connectionName), 1);
        }

        public LinqToDbSettings(ConnectionStringSettings defaultProvider, params ConnectionStringSettings[] additionalProviders)
        {
            DefaultConfiguration = defaultProvider.ProviderName;
            DefaultDataProvider = defaultProvider.ProviderName;
            ConnectionStrings = additionalProviders.Prepend(defaultProvider);
        }
    }


```

**Comment by MaceWindu** [comment, marked as answer]
You can register connection interceptor, which will setup connection on open for specific context

https://linq2db.github.io/articles/general/interceptors.html#iconnectioninterceptor

https://linq2db.github.io/articles/general/interceptors.html#interceptors-registration

**Comment by ademchenko** [comment]
@MaceWindu , is it possible if cli tool will also generate that mapping along with the entities? Because logically that mapping is highly tied to the generated entities - it's used there but we need to create an interceptor, register that, etc. etc to get enums worked. It looks overcomplicated.

**Reply by MaceWindu** [reply]
It's tracked by #3501
